### PR TITLE
Document warning waves.

### LINKED ISSUE #20355: Document the difference between WarningLevel and AnalysisLevel
The documentation for -warn compiler option should mention notes about AnalysisLevel, is it the same as WarningLevel? is it just a new name for it?

https://devblogs.microsoft.com/dotnet/automatically-find-latent-bugs-in-your-code-with-net-5/

This blog mentions that the default value for AnalysisLevel is 5. However, the default value for WarningLevel is 4.

All these need to be clarified.

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 26ef35c2-6df4-26f8-32b2-9b50eb76c037
* Version Independent ID: e57ac82b-307e-163b-8172-52644b3a6baa
* Content: [C# Compiler Options - errors and warnings](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/compiler-options/errors-warnings#warninglevel)
* Content Source: [docs/csharp/language-reference/compiler-options/errors-warnings.md](https://github.com/dotnet/docs/blob/main/docs/csharp/language-reference/compiler-options/errors-warnings.md)
* Product: **dotnet-csharp**
* Technology: **csharp-language-reference**
* GitHub Login: @BillWagner
* Microsoft Alias: **wiwagn**

### LINKED ISSUE #24089: Enhance ReportAnalyzer docs
I never got it to work without binarylog. Is there a way to get it to work without binarylog? If there isn't, the docs must mention that it should be used combined with bl flag.

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 26ef35c2-6df4-26f8-32b2-9b50eb76c037
* Version Independent ID: e57ac82b-307e-163b-8172-52644b3a6baa
* Content: [C# Compiler Options - errors and warnings](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/compiler-options/errors-warnings)
* Content Source: [docs/csharp/language-reference/compiler-options/errors-warnings.md](https://github.com/dotnet/docs/blob/main/docs/csharp/language-reference/compiler-options/errors-warnings.md)
* Product: **dotnet-csharp**
* Technology: **csharp-language-reference**
* GitHub Login: @BillWagner
* Microsoft Alias: **wiwagn**

### LINKED ISSUE #26199: TreatWarningsAsErrors does not behave like /warnaserror
The page implies that these are two syntax forms to do the same thing, when in fact they differ in at least 2 regards:
1. `/warnaserror` promotes ALL msbuild warnings to errors, whereas `TreatWarningsAsErrors` only affects compiler (`csc`) warnings 
2. `/warnaserror` generates output even in the face of errors, whereas `TreatWarningsAsErrors` does not

More details: https://stackoverflow.com/questions/54771043/treatwarningsaserrors-vs-warnaserror

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 26ef35c2-6df4-26f8-32b2-9b50eb76c037
* Version Independent ID: e57ac82b-307e-163b-8172-52644b3a6baa
* Content: [C# Compiler Options - errors and warnings](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/compiler-options/errors-warnings)
* Content Source: [docs/csharp/language-reference/compiler-options/errors-warnings.md](https://github.com/dotnet/docs/blob/main/docs/csharp/language-reference/compiler-options/errors-warnings.md)
* Product: **dotnet-csharp**
* Technology: **csharp-language-reference**
* GitHub Login: @BillWagner
* Microsoft Alias: **wiwagn**

### LINKED ISSUE #27661: Unclear which syntax to use to list warnings
On this page, we have a few examples: `<WarningsAsErrors>0219,0168</WarningsAsErrors>`, `<WarningsNotAsErrors>0219,0168</WarningsNotAsErrors>` and `<DisabledWarnings>number1, number2</DisabledWarnings>`.

They are only showing numbers, while some of the warnings displayed may be `NU...`, `CS...`, or even `MSB...`.

How are we supposed to list these warnings?
Is the syntax `<WarningsNotAsErrors>NU1603</WarningsNotAsErrors>` supposed to work?

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 26ef35c2-6df4-26f8-32b2-9b50eb76c037
* Version Independent ID: e57ac82b-307e-163b-8172-52644b3a6baa
* Content: [C# Compiler Options - errors and warnings](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/compiler-options/errors-warnings)
* Content Source: [docs/csharp/language-reference/compiler-options/errors-warnings.md](https://github.com/dotnet/docs/blob/main/docs/csharp/language-reference/compiler-options/errors-warnings.md)
* Product: **dotnet-csharp**
* Technology: **csharp-language-reference**
* GitHub Login: @BillWagner
* Microsoft Alias: **wiwagn**

### LINKED ISSUE #27872: SARIF version in C# compiler ErrorLog property
Please document how to specify the version of the SARIF format as part of the ErrorLog property used by the C# compiler. This feature was added in <https://github.com/dotnet/roslyn/pull/37148>.

```XML
<ErrorLog>log21.json,version=2.1</ErrorLog>
```

The compiler accepts either a comma or a semicolon as the separator between the file name and `version`, but its `--help` output uses a comma.

If the ErrorLog property is set in the `dotnet build` command line, then the comma (or semicolon) must be percent-encoded as in `dotnet build -p:ErrorLog=logcmd.json%2Cversion=2.1`, to prevent the `-p` option from treating `version` as the name of a separate property.

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 26ef35c2-6df4-26f8-32b2-9b50eb76c037
* Version Independent ID: e57ac82b-307e-163b-8172-52644b3a6baa
* Content: [C# Compiler Options - errors and warnings](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/compiler-options/errors-warnings#errorlog)
* Content Source: [docs/csharp/language-reference/compiler-options/errors-warnings.md](https://github.com/dotnet/docs/blob/main/docs/csharp/language-reference/compiler-options/errors-warnings.md)
* Product: **dotnet-csharp**
* Technology: **csharp-language-reference**
* GitHub Login: @BillWagner
* Microsoft Alias: **wiwagn**

### LINKED ISSUE #27927: Warning level documentation is outdated
> Any value greater than 5 will be treated as 5. To make sure you always have all warnings if the compiler is updated with new warning levels, put an arbitrary large value (for example, 9999).

This is not correct. We have warning level 7 now.

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 26ef35c2-6df4-26f8-32b2-9b50eb76c037
* Version Independent ID: e57ac82b-307e-163b-8172-52644b3a6baa
* Content: [C# Compiler Options - errors and warnings](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/compiler-options/errors-warnings#warninglevel)
* Content Source: [docs/csharp/language-reference/compiler-options/errors-warnings.md](https://github.com/dotnet/docs/blob/main/docs/csharp/language-reference/compiler-options/errors-warnings.md)
* Product: **dotnet-csharp**
* Technology: **csharp-language-reference**
* GitHub Login: @BillWagner
* Microsoft Alias: **wiwagn**

### LINKED ISSUE #28806: Set of WarningLevel values is incomplete
The set of WarningLevel values is incomplete, and the description for 5 is also incomplete. We want something like:

- 5: Includes warnings from C# 9.0 and built-in .NET 5.0 analyzers.
- 6: Includes warnings from C# 10.0 and built-in .NET 6.0 analyzers.
- 9999: Includes all warnings.

These are all values the .NET 6.0 SDK itself sets in Microsoft.NET.Sdk.Analyzers.targets.

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 26ef35c2-6df4-26f8-32b2-9b50eb76c037
* Version Independent ID: e57ac82b-307e-163b-8172-52644b3a6baa
* Content: [C# Compiler Options - errors and warnings](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/compiler-options/errors-warnings)
* Content Source: [docs/csharp/language-reference/compiler-options/errors-warnings.md](https://github.com/dotnet/docs/blob/main/docs/csharp/language-reference/compiler-options/errors-warnings.md)
* Product: **dotnet-csharp**
* Technology: **csharp-language-reference**
* GitHub Login: @BillWagner
* Microsoft Alias: **wiwagn**

### LINKED ISSUE #29265: [17.3] Warning wave on lower case type names.
lower case type names will emit a warning, if the warning level is 7 or higher.

See [feature spec](https://github.com/dotnet/roslyn/blob/main/docs/compilers/CSharp/Warnversion%20Warning%20Waves.md#warning-level-7) for details.

### PR DESCRIPTION
**Pull request by BillWagner**
Document the warning wave warnings from C# 9 through C# 11 (as of preview 17.3).

While editing the page on compiler warnings and errors, fix all remaining open issues.

- [Warning waves](https://review.docs.microsoft.com/en-us/dotnet/csharp/language-reference/compiler-messages/warning-waves?branch=pr-en-us-29283) internal review link.
- [WarningLevel](https://review.docs.microsoft.com/en-us/dotnet/csharp/language-reference/compiler-options/errors-warnings?branch=pr-en-us-29283#warninglevel) updates internal review link.

Fixes #20355
Fixes #24089
Fixes #26199
Fixes #27661
Fixes #27872
Fixes #27927
Fixes #28806 
Fixes #29265

### CODE REVIEW

File: docs/csharp/language-reference/compiler-messages/warning-waves.md [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+
+The following examples demonstrate this warning. The expression binds incorrectly because of the precedence of the operators.
+
+:::code language="csharp" source="./snippets/WarningWaves/WaveFive.cs" id="QueryPrecedence":::
+
+To fix this error, put parentheses around the query expression:
+
+:::code language="csharp" source="./snippets/WarningWaves/WaveFive.cs" id="QueryPrecedenceNoWarn":::
+
+## Members must be fully assigned, use of unassigned variable (CS8880, CS8881, CS8882, CS8883, CS8884, CS8885, CS8886, CS8887)
```
Comment (BillWagner) [inline docs/csharp/language-reference/compiler-messages/warning-waves.md:107]: @tdykstra 

This section was very redundant if I had a separate header for each warning. In this commit, I tried to put some key words shared among the messages and added all the IDs. I'm ambivalent about that format. What do you think? Do you have any better suggestion here?

File: docs/csharp/language-reference/compiler-messages/warning-waves.md [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+
+The following examples show the warnings generated from the improved definite assignment analysis:
+
+- CS8880:  Auto-implemented property 'Property' must be fully assigned before control is returned to the caller.
+- CS8881:  Field 'field' must be fully assigned before control is returned to the caller.
+- CS8882: The out parameter 'parameter' must be assigned to before control leaves the current method.
+- CS8883: Use of possibly unassigned auto-implemented property 'Property'.
+- CS8884: Use of possibly unassigned field 'Field'
+- CS8885: The 'this' object can't be used before all its fields have been assigned.
+- CS8886: Use of unassigned output parameter `parameterName`
```
Comment (tdykstra) [inline docs/csharp/language-reference/compiler-messages/warning-waves.md:123]:
```suggestion
- CS8886: Use of unassigned output parameter 'parameterName'.
```

File: docs/csharp/language-reference/compiler-messages/warning-waves.md [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+The following example generates CS8892:
+
+:::code language="csharp" source="./snippets/WarningWaves/Program.cs" id="MultipleEntryPoints":::
+
+> [!NOTE]
+> The compiler always uses the synchronous entry point. In case there are multiple synchronous entry points, you get a compiler error.
+
+To fix this warning, remove or rename the asynchronous entry point.
+
+## CS8897 - static types can't be used as parameters
```
Comment (tdykstra) [inline docs/csharp/language-reference/compiler-messages/warning-waves.md:147]:
```suggestion
## CS8897 - Static types can't be used as parameters
```

File: docs/csharp/language-reference/compiler-options/errors-warnings.md [hunk 4]
```diff
@@ -37,17 +38,31 @@ The **WarningLevel** option specifies the warning level for the compiler to disp
...
+| 2 | Displays level 1 warnings plus certain, less-severe warnings, such as warnings about hiding class members. |
+| 3 | Displays level 2 warnings plus certain, less-severe warnings, such as warnings about expressions that always evaluate to `true` or `false`. |
+| 4   (the default)|Displays all level 3 warnings plus informational warnings. |
 
-> [!NOTE]
-> To make sure you always have all warnings if the compiler is updated with new warning levels, put an arbitrary large value (for example, `9999`).
+> [!WARNING]
+> The compiler command line accepts values greater than 4 to enable [warning wave warnings](../compiler-messages/warning-waves.md). However, the .NET SDK sets the *WarningLevel* to match the *AnalysisLevel* in your project file.
+
+To get information about an error or warning, you can look up the error code in the Help Index. For other ways to get information about an error or warning, see [C# Compiler Errors](../compiler-messages/index.md). Use [**TreatWarningsAsErrors**](#treatwarningsaserrors) to treat all warnings as errors. Use [**DisabledWarnings**](#disabledwarnings) to disable certain warnings.
```
Comment (tdykstra) [inline docs/csharp/language-reference/compiler-options/errors-warnings.md:52]: What is the Help Index, is it something we can link to?

File: docs/csharp/language-reference/compiler-options/errors-warnings.md [hunk 5]
```diff
@@ -37,17 +38,31 @@ The **WarningLevel** option specifies the warning level for the compiler to disp
...
+| Analysis level   | Meaning |
+|------------------|---------|
+| 5                | Displays all optional [warning wave 5 warnings](../compiler-messages/warning-waves.md#cs7023---a-static-type-is-used-in-an-is-or-as-expression). |
+| 6                | Displays all optional [warning wave 6 warnings](../compiler-messages/warning-waves.md#cs8826---partial-method-declarations-have-signature-differences). |
+| 7                | Displays all optional [warning wave 7 warnings](../compiler-messages/warning-waves.md#cs8981---the-type-name-only-contains-lower-cased-ascii-character). |
+| latest (default) | Displays all informational warnings up to and including the current release. |
+| preview          | Displays all informational warnings up to and including the latest preview release. |
+| none             | Turns off all informational warnings. |
+
+For more information on optional warnings, see [Warning waves](../compiler-messages/warning-waves.md).|
```
Comment (tdykstra) [inline docs/csharp/language-reference/compiler-options/errors-warnings.md:65]:
```suggestion
For more information on optional warnings, see [Warning waves](../compiler-messages/warning-waves.md).
```

File: docs/csharp/language-reference/compiler-options/errors-warnings.md [hunk 6]
```diff
@@ -115,6 +133,14 @@ Specify a file to log all compiler and analyzer diagnostics.
 
 The **ErrorLog** option causes the compiler to output a [Static Analysis Results Interchange Format (SARIF) log](https://github.com/microsoft/sarif-tutorials/blob/main/docs/1-Introduction.md#:~:text=What%20is%20SARIF%3F,for%20use%20by%20simpler%20tools). SARIF logs are typically read by tools that analyze the results from compiler and analyzer diagnostics.
 
+You can specify the SARIF format using the `version` argument to the `ErrorLog` element:
+
+```XML
+<ErrorLog>logVersion21.json,version=2.1</ErrorLog>
+```
+
+The separator can be either a comma (`,`) or a semi-colon (`;`). Valid values are for version are:  "1", "2", and "2.1". The default is "1". "2" and "2.1" both mean SARIF version 2.1.0.
```
Comment (tdykstra) [inline docs/csharp/language-reference/compiler-options/errors-warnings.md:142]:
```suggestion
The separator can be either a comma (`,`) or a semicolon (`;`). Valid values are for version are:  "1", "2", and "2.1". The default is "1". "2" and "2.1" both mean SARIF version 2.1.0.
```

**Review by tdykstra — APPROVED** [review]
Looks great, I noted just a few formatting suggestions.

File: docs/csharp/language-reference/compiler-options/errors-warnings.md [hunk 7]
```diff
@@ -19,6 +19,7 @@ helpviewer_keywords:
 The following options control how the compiler reports errors and warnings. The new MSBuild syntax is shown in **Bold**. The older *csc.exe* syntax is shown in `code style`.
 
 - **WarningLevel** / `-warn`: Set warning level.
+- **AnalysisLevel**: Set optional warning level.
```
[the remark this answers is not in the document]
Comment (BillWagner) [reply docs/csharp/language-reference/compiler-options/errors-warnings.md:22]: Yes. It doesn't have a command line switch equivalent.

File: docs/csharp/language-reference/compiler-messages/warning-waves.md (hunk 1 above)
Comment (tdykstra) [reply docs/csharp/language-reference/compiler-messages/warning-waves.md:107]: This makes sense to me, though I'm ambivalent about it too, as it makes for a very long heading.  The only revision I might consider is to put the warning numbers first to match the sequence in the other headings. You could make it a range --

CS8880-8887

-- but I don't know if that would compromise SEO for 81-87

File: docs/csharp/language-reference/compiler-options/errors-warnings.md [hunk 8]
```diff
@@ -61,20 +76,23 @@ The **TreatWarningsAsErrors** option treats all warnings as errors. You can also
...
 
+> [!IMPORTANT]
+> There are two subtle differences between using the `<TreatWarningsAsErrors>` element in your *csproj* file, and using the `warnaserror` MSBuild command line switch. *TreatWarningsAsErrors* only impacts the C# compiler, not any other MSBuild tasks in your *csproj* file. The `warnaserror` command line switch impacts all tasks. Secondly, the compiler doesn't produce any output on any warnings when *TreatWarningsAsErrors* is used. The compiler produces output when the `warnaserror` command line switch is used.
+
 ## WarningsAsErrors and WarningsNotAsErrors
 
-The **WarningsAsErrors** and **WarningsNotAsErrors** options override the **TreatWarningsAsErrors** option for a list of warnings.
+The **WarningsAsErrors** and **WarningsNotAsErrors** options override the **TreatWarningsAsErrors** option for a list of warnings. This option can be used with all *CS* warnings. The "CS" prefix is optional. You can use either the number, or "CS" followed by the error or warning level. For other elements that affect warnings, see the [Common MSBuild properties](/visualstudio/msbuild/common-msbuild-project-properties).
```
Comment (KalleOlaviNiemitalo) [inline docs/csharp/language-reference/compiler-options/errors-warnings.md:84]: Is it correct that this says "warning level" and not "warning number"? IIRC, warning levels are 1 to 4, so does this mean e.g. "CS3" would affect all level-3 warnings?

File: docs/csharp/language-reference/compiler-options/errors-warnings.md [hunk 9]
```diff
@@ -115,6 +133,14 @@ Specify a file to log all compiler and analyzer diagnostics.
 
 The **ErrorLog** option causes the compiler to output a [Static Analysis Results Interchange Format (SARIF) log](https://github.com/microsoft/sarif-tutorials/blob/main/docs/1-Introduction.md#:~:text=What%20is%20SARIF%3F,for%20use%20by%20simpler%20tools). SARIF logs are typically read by tools that analyze the results from compiler and analyzer diagnostics.
 
+You can specify the SARIF format using the `version` argument to the `ErrorLog` element:
+
+```XML
+<ErrorLog>logVersion21.json,version=2.1</ErrorLog>
+```
+
+The separator can be either a comma (`,`) or a semicolon (`;`). Valid values are for version are:  "1", "2", and "2.1". The default is "1". "2" and "2.1" both mean SARIF version 2.1.0.
```
Comment (KalleOlaviNiemitalo) [inline docs/csharp/language-reference/compiler-options/errors-warnings.md:142]: Double "are".

File: docs/csharp/language-reference/compiler-options/errors-warnings.md [hunk 10]
```diff
@@ -124,3 +150,6 @@ Report additional analyzer information, such as execution time.
 ```
 
 The **ReportAnalyzer** option causes the compiler to emit extra MSBuild log information that details the performance characteristics of analyzers in the build. It's typically used by analyzer authors as part of validating the analyzer.
+
+> [!IMPORTANT]
+> The extra log information generated by this flag is only generated when the `-verbosity:detailed` command line option is used. See the [swtiches](/visualstudio/msbuild/msbuild-command-line-reference#switches) article in the MSBuild documentation for more information.
```
Comment (KalleOlaviNiemitalo) [inline docs/csharp/language-reference/compiler-options/errors-warnings.md:155]: Typo "swtiches".

File: docs/csharp/language-reference/compiler-messages/warning-waves.md (hunk 1 above)
Comment (BillWagner) [reply docs/csharp/language-reference/compiler-messages/warning-waves.md:107]: Good point @tdykstra 

For now, I'll leave it as merged. I'll check stats and make any updates next month. I'll see how many searches we get on the warning code, vs. direct traffic from the F1 service.

File: docs/csharp/language-reference/compiler-options/errors-warnings.md (hunk 9 above)
Comment (BillWagner) [reply docs/csharp/language-reference/compiler-options/errors-warnings.md:142]: addressed in #29435

**Pull request merged**